reducer.py

```diff
@@ -0,0 +1,319 @@
+        self.fetches = fetches
+        self.reducer = reducer
+        self.metric_name = metric_name
+
+    def begin(self):
+        self.batch_metrics = []
+
+    def before_run(self, run_contex):
+        return tf.estimator.SessionRunArgs(fetches=self.fetches)
+
+    def after_run(self, run_contex, run_values):
+        self.per_slot_metrics.append(run_values.results)
+
+    def end(self, session):
+        reduced = self.context.allreduce_metrics(self.reducer, self.batch_metrics)
+
+        write_metric_to_tensorboard(self.metric_name, reducer)
+
+
+class MyLinearEstimator(estimator.EstimatorTrial):
+    ...
+
+    def build_validation_spec(self) -> tf.estimator.EvalSpec:
+        def fn():
+            ...
+
+        global_step_metric_hook = det.estimator.ReducerEvalHook(
+            context=self.context,
+            fetches="my_variable:0",
+            reducer=MyReducer(),
+            metric_name="my_metric",
+        )
+
+        return tf.estimator.EvalSpec(fn, hooks=[global_step_metric_hook])
+
+        # Shiyuan says: maybe make this a separate callback and hide the ReducerEvalHook thing?
+
+
+
+# How would we integrate with PyTorchTrial?
+#
+# Basically exactly how we currently support selecting reducers, only we would accept instances
+# of Reducer instead of the det.pytorch.Reducer enum (we'd could easily be backwards
+# compatible though):
+
+class MyPytorchTrial(PyTorchTrial):
+    ...
+
+    def evaluation_reducer(self):
+        return {"error": MyErrorReducer(), "accuracy": MyAccuracyReducer()}
+
+
+
+
+#######################
+# Alternate idea: it seems like the Reducer interface described above is simple, but it enforces
+# poor memory usage patterns.  What if we added an accumulate() method?  Things like confusion
+# matrices could be accumualated in very memory-efficient ways.
+
+class Reducer:
+    """
+    A two-stage hierarchical reduction.  For a simpler interface, see SimpleReducer, below.
+
+    Unlike the current internal reducers, both steps will always run, and users who want a simpler
+    interface will let the first step be an identity function.
+    """
+    # TODO: add a Reducer.reset() method, or just build a new Reducer instance every cycle?
+
+    @abc.abstractmethod
+    def accumulate(self, metric):
+        """
+        Accumulate a metric.  For simple reducers like sums or means, you might just add the metric
+        to an accumulator.  For complex reducers, you may have to store the original values in
+        memory.
+        """
+        pass
+
+    @abc.abstractmethod
+    def per_slot_reduce(self, metrics):
+        """
+        d-train optimization: do the bulk of your reduction on each GPU.
+
+        metrics is literally just a list of the metrics your trial returned, no collation at all.
+
+        This is always called, in single- and multi-gpu settings.
+        """
+        pass
+
+    @abc.abstractmethod
+    def cross_slot_reduce(self, per_slot_metrics):
```

Comment:
True, but why not make it all-gather semantics?  It seems like the communication overhead will be effectively zero here, and it can only make writing user code easier, like for deciding on early stopping after reaching a certain metric or something.